near-sdk/src/collections/tree_map.rs

```diff
@@ -0,0 +1,1300 @@
+    // enforcing balance (if imbalance takes place) along the way up to the root.
+    fn check_balance(&mut self, at: u64, key: &K) -> u64 {
+        match self.key.get(&at) {
+            Some(k) => {
+                if k.eq(key) {
+                    self.update_height(at);
+                    at
+                } else {
+                    if k.lt(key) {
+                        match self.lft.get(&at) {
+                            Some(l) => {
+                                let id = self.check_balance(l, key);
+                                self.lft.insert(&at, &id);
+                            },
+                            None => ()
+                        }
+                    } else {
+                        match self.rgt.get(&at) {
+                            Some(r) => {
+                                let id = self.check_balance(r, key);
+                                self.rgt.insert(&at, &id);
+                            },
+                            None => ()
+                        }
+                    }
+                    self.update_height(at);
+                    self.enforce_balance(at)
+                }
+            },
+            None => at
+        }
+    }
+
+    // Node holding the key is not removed from the tree - instead the substitute node is found,
+    // the key is copied to 'removed' node from substitute node, and then substitute node gets
+    // removed from the tree.
+    //
+    // The substitute node is either:
+    // - right-most (max) node of the left subtree (containing smaller keys) of node holding `key`
+    // - or left-most (min) node of the right subtree (containing larger keys) of node holding `key`
+    //
+    fn do_remove(&mut self, key: &K) -> u64 {
+        let root = self.root;
+        // r_id - id of a node containing key of interest
+        // r_p - id of an immediate parent node of r_id
+        let (r_id, r_p) = match self.lookup_at(root, key) {
+            Some(x) => x,
+            None => return root // cannot remove a missing key, no changes to the tree needed
+        };
+
+        let lft_opt = self.lft.get(&r_id);
+        let rgt_opt = self.rgt.get(&r_id);
+
+        if lft_opt.is_none() && rgt_opt.is_none() {
+            // remove leaf
+            let p_key = self.key.get(&r_p).unwrap();
+            if p_key.lt(key) {
+                self.rgt.remove(&r_p);
+            } else {
+                self.lft.remove(&r_p);
+            }
+            self.key.remove(&r_id);
+            self.ht.remove(&r_id);
+
+            // removing node might have caused a imbalance - balance the tree up to the root,
+            // starting from lowest affected key - the parent of a leaf node in this case
+            self.check_balance(root, &p_key)
+
+        } else {
+            // non-leaf node, select subtree to proceed with
+            let b = self.get_balance(r_id);
+            if (b >= 0 && lft_opt.is_some()) || rgt_opt.is_none() {
```

Comment:
It should be enough to check `b >= 0` right? If we know that at least one child is set, and the left tree is no smaller than the right tree, than the left tree must exist.
Similarly, if `b < 0` rgt_opt will never be `None`

Reply:
done